### ISSUE TITLE
toml.Unmarshal overrides origin values if key is omitted in the file

### ISSUE THREAD

**Issue by troian** [pelletier/go-toml#252]
**Describe the bug**
When unmarshaling to struct that already have some default value toml.Unmarshal overwrites it
when key is not present in the file

**To Reproduce**
Steps to reproduce the behavior. Including TOML files.
```go

package main

import(
	"github.com/pelletier/go-toml"
)

type config struct {
	Val1 string `toml:"val1"`
	Val2 string `toml:"val2"`
}

var configFile = []byte(
	`
val1 = "test1"
`)

func main() {
	cfg := &config {
		Val2: "test2",
	}

	err := toml.Unmarshal(configFile, cfg)
	if err != nil {
		panic(err.Error())
	}

	if cfg.Val2 != "test2" {
		panic("Val2 overwritten")
	}
}
```

**Expected behavior**
Should remain default value if no key is present in the file

**Versions**
 - go-toml: b56e1b2
 - go: version 1.11.2
 - operating system: macOS

**Comment by fopina** [comment]
👍 just like json in stdlib https://play.golang.org/p/-iiZnh5uhy

**Comment by moorereason** [comment]
@fopina,
Wrong.  `B` is retained by the `encoding/json` package.

**Comment by moorereason** [comment]
I apologize.  I misunderstood your comment.  I thought you meant that go-toml was acting like the stdlib.

**Comment by pelletier** [comment]
I took a quick stab at it. Not as easy as I thought it would be. The root cause is that when unmarshalling, [the whole object is replaced](https://github.com/pelletier/go-toml/blob/master/marshal.go#L462.), instead of being merged. My long-term plan to rewrite the parser inside-out to build the unmarshalled result as it runs would fix it, but just can't find the time to work on it.

In the meantime, one option would be to turn that `Set` operation into a merge operation when processing structs and maps. Not elegant but should get the job done.

**Comment by troian** [comment]
Btw, another "issue" (I'm not sure it's actually issue) but might be related to parser as well
When marshaling object to stream keys order are getting into completely reordered. Is that related too?

**Comment by pelletier** [comment]
This is actually the expected behavior since https://github.com/pelletier/go-toml/pull/133. This is a workaround the fact that the marshaling process is backed by TomlTree, which stores everything in maps. Those maps do not have a guaranteed order. One of the very common use-case of go-toml is parse -> edit tree -> write. To make that process easier, we decided to introduce keys ordering when serializing TomlTrees. Ideally, the marshaling process should respect the field order of the struct.

That issue is not related to this bug.

**Comment by troian** [comment]
Is it ok to make another one then? For example our projects marshals structs to generate configs and as result it looks like complete mess)

**Comment by pelletier** [comment]
How do you expect the `default:"..."` tag to work?

* Option 1: it takes precedence of already existing value
* Option 2: existing value takes precedence over default tag

```go
type One struct {
	Field string `default:"str default"`
}

o := &One {
	Field: "value",
}

toml.Unmarshal([]byte(``), o)

// option 1: o.Field == "str default"
// option 2: o.Field == "value"
```

Option 2 may be tricky to implement for some types. For example, string's "default" is `""`. So if an empty string is provided in a field, do you consider it being a user-provided value (therefore, the value stays an empty string), or is it just its init value, (so the `default` tag's value be used instead)?

**Comment by dahu33** [comment]
IMO, the default tag should be deprecated in favor of user-provided values (which supports all types while the `default` tag doesn't). During the deprecation period, option 1 (take precedence of already existing value) seems the best option for the following reason:
1) to preserve compatibility with older code
2) ease of implementation and maintenance

Just my 2cents :)

**Comment by pelletier** [comment]
After two and a half years, this is fixed in the upcoming v2!

**Comment by troian** [comment]
shall issue be closed then?

**Comment by pelletier** [comment]
I’ll close all the issues with the tag “v2-fixed” when v2.0.0 is released.

**Comment by pelletier** [comment]
Closing this issue, as [go-toml v2.0.0](https://github.com/pelletier/go-toml/releases/tag/v2.0.0) has been released and does not have this problem. As go-toml v1 will not be receiving any updates, please look into [upgrading to the new version](https://github.com/pelletier/go-toml/#migrating-from-v1).
